tough/src/editor.rs

```diff
@@ -0,0 +1,660 @@
+        rng: &dyn SecureRandom,
+    ) -> Result<SignedRole<T>>
+    where
+        T: Role + Serialize,
+    {
+        let root = &self.signed_root.signed.signed;
+        let role_keys = root.roles.get(&T::TYPE).context(error::NoRoleKeysinRoot {
+            role: &T::TYPE.to_string(),
+        })?;
+        let (signing_key_id, signing_key) = keys
+            .iter()
+            .find(|(keyid, _signing_key)| role_keys.keyids.contains(&keyid))
+            .context(error::SigningKeysNotFound {
+                role: &T::TYPE.to_string(),
+            })?;
+
+        let mut role = Signed {
+            signed: role,
+            signatures: Vec::new(),
+        };
+
+        let mut data = Vec::new();
+        let mut ser = serde_json::Serializer::with_formatter(&mut data, CanonicalFormatter::new());
+        role.signed
+            .serialize(&mut ser)
+            .context(error::SerializeRole)?;
+        let sig = signing_key.sign(&data, rng)?;
+        role.signatures.push(Signature {
+            keyid: signing_key_id.clone(),
+            sig: sig.into(),
+        });
+
+        let mut buffer = serde_json::to_vec_pretty(&role).context(error::SerializeSignedRole)?;
+        buffer.push(b'\n');
+        let length = buffer.len() as u64;
+
+        let mut sha256 = [0; SHA256_OUTPUT_LEN];
+        sha256.copy_from_slice(digest(&SHA256, &buffer).as_ref());
+
+        let signed_role = SignedRole {
+            signed: role,
+            buffer,
+            sha256,
+            length,
+        };
+
+        Ok(signed_role)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::path::PathBuf;
+
+    fn root_path() -> PathBuf {
+        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .join("tests")
+            .join("data")
+            .join("tuf-reference-impl")
+            .join("metadata")
+            .join("root.json")
+    }
+
+    // Make sure we can add existing role structs and the proper data is kept.
+    #[test]
+    fn existing_roles() {
+        let targets: Signed<Targets> = serde_json::from_str(include_str!(
+            "../tests/data/tuf-reference-impl/metadata/targets.json"
+        ))
+        .unwrap();
+        let snapshot: Signed<Snapshot> = serde_json::from_str(include_str!(
+            "../tests/data/tuf-reference-impl/metadata/snapshot.json"
+        ))
+        .unwrap();
+        let timestamp: Signed<Timestamp> = serde_json::from_str(include_str!(
+            "../tests/data/tuf-reference-impl/metadata/timestamp.json"
+        ))
+        .unwrap();
+        let root_path = root_path();
+
+        let mut editor = RepositoryEditor::new(&root_path).unwrap();
+        editor
+            .targets(targets.signed)
+            .unwrap()
+            .snapshot(snapshot.signed)
+            .unwrap()
+            .timestamp(timestamp.signed)
+            .unwrap();
+
+        let targets = editor.existing_targets.as_ref().unwrap();
+
+        assert!(editor.targets_version.is_none());
```

Comment:
I'm not sure of the end goal of this test function, other than the unwraps not failing - these `is_none` checks don't seem to be helping us much since that'd be the default :)  I guess it checks that file1/file2 are in targets?  Is that enough to have confidence?

Reply:
The issue might be that we are using the simplistic 'reference implementation' tuf repo. We should consider adding a more complex tuf repo to the test data. One that has consistent snapshots and multiple keys, for example. In other words, the assertions are as boring as the tuf repo is.

Reply:
I guess I saw this as more of a "unit" test for the methods that add data from existing roles. I just want to make sure that we only keep the data we want to.